### PR TITLE
[Go SDK]: Implement fileio.MatchContinuously transform

### LINKED ISSUE #26186: [Feature Request][Go SDK]: Add fileio.MatchContinuously transform
### What would you like to happen?

Add a `fileio.MatchContinuously` transform which takes a glob pattern and continuously outputs matching files

### Issue Priority

Priority: 2 (default / most feature requests should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [X] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by johannaojeling**
Fixes #26186 and adds a `fileio.MatchContinuously` transform to the Go SDK.

I have tested the transform manually on Dataflow and Flink, which seem to be the only runners that currently support all of watermark estimation, self-checkpointing and state provisioning, that this transform relies on. I have not added any unit tests for the same reason, and there were limitations with running integration tests with some of these features for Dataflow and Flink so I have skipped that as well.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/go/pkg/beam/io/fileio/match.go
Comment (johannaojeling) [inline sdks/go/pkg/beam/io/fileio/match.go:265]: I would like to also enable emitting a file only if it has been modified since the last observation, but it requires introduction of a new method to the filesystems for retrieving a file's last modified timestamp first (#26187) + attaching this info to the `fileio.FileMetadata` in `fileio.matchFn`, so I will do it at a later stage

**Review by riteshghorse — APPROVED** [review]
This is great, thanks for your contribution!

**Comment by lostluck** [comment]
Ah foiled! We can fix this a bit once state is added to Prism, since that's what's currently missing here.

File: sdks/go/pkg/beam/io/fileio/match.go
Comment (lostluck) [reply sdks/go/pkg/beam/io/fileio/match.go:265]: I'll review PR 26192 first then.

**Comment by lostluck** [comment]
Happy to merge this in if there are no further question/disccusion (after one of us fixes the conflict in CHANGES.md of course!)

**Review by lostluck — APPROVED** [review]
LGTM. Happy to merge this in, unless you want to have the files modified bit come in first/same time. Let me know, and I'll merge as appropriate.

I don't know if anything from the Java/Python equivalents require timers though, which are almost ready to go in.

**Comment by johannaojeling** [comment]
Thank you both for reviewing!

The Java and Python equivalents don't use timers. Happy too add tests for this one later on when state is supported.

I can open a separate PR for the modified files part once #26192 has been merged so please go ahead and merge.

**Review by lostluck — APPROVED** [review]
SGTM. I'm probably over-indexing on the importance of garbage collecting state. It's unlikely to be a problem for any but the most egregious number of files.

**Pull request merged**
